src/Workspaces/CoreTest/UtilityTest/SpecializedTasksTests.cs

```diff
@@ -63,5 +67,490 @@ public void WhenAll_NotYetCompleted()
+            var gate = new ManualResetEventSlim();
+            Func<StateType, CancellationToken, ValueTask<IntermediateType>> func = async (_, _) =>
+            {
+                await Task.Yield();
+                gate.Wait(CancellationToken.None);
+                throw fault;
+            };
+            Func<IntermediateType, StateType, ResultType> transform = (_, _) =>
+            {
+                executedTransform = true;
+                return new ResultType();
+            };
+            var arg = new StateType();
+
+            var task = SpecializedTasks.TransformWithoutIntermediateCancellationExceptionAsync(func, transform, arg, cancellationToken).Preserve();
+            Assert.False(task.IsCompleted);
+
+            gate.Set();
+            var exception = await Assert.ThrowsAsync<TaskCanceledException>(() => task.AsTask());
+            Assert.True(task.IsCanceled);
+            Assert.Equal(cancellationToken, exception.CancellationToken);
+            Assert.False(executedTransform);
+        }
+
+        [Fact]
+        public void Transform_SyncCompletedFunction_FaultedTransform()
+        {
+            var fault = ExceptionUtilities.Unreachable;
+            Func<StateType, CancellationToken, ValueTask<IntermediateType>> func = (_, _) => new(new IntermediateType());
+            Func<IntermediateType, StateType, ResultType> transform = (_, _) => throw fault;
+            var arg = new StateType();
+            var cancellationToken = new CancellationToken(canceled: false);
+
+#pragma warning disable CA2012 // Use ValueTasks correctly (the instance is never created)
+            var exception = Assert.Throws<InvalidOperationException>(() => SpecializedTasks.TransformWithoutIntermediateCancellationExceptionAsync(func, transform, arg, cancellationToken));
+#pragma warning restore CA2012 // Use ValueTasks correctly
+            Assert.Same(fault, exception);
+        }
+
+        [Fact]
+        public async Task Transform_AsyncCompletedFunction_FaultedTransform()
+        {
+            var fault = ExceptionUtilities.Unreachable;
+            var gate = new ManualResetEventSlim();
+            Func<StateType, CancellationToken, ValueTask<IntermediateType>> func = async (_, _) =>
+            {
+                await Task.Yield();
+                gate.Wait(CancellationToken.None);
+                return new IntermediateType();
+            };
+            Func<IntermediateType, StateType, ResultType> transform = (_, _) => throw fault;
+            var arg = new StateType();
+            var cancellationToken = new CancellationToken(canceled: false);
+
+            var task = SpecializedTasks.TransformWithoutIntermediateCancellationExceptionAsync(func, transform, arg, cancellationToken).Preserve();
+            Assert.False(task.IsCompleted);
+
+            gate.Set();
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(() => task.AsTask());
+            Assert.Same(fault, exception);
```

Comment:
i skimmed.  but i assume you're testing the table below?

Reply:
It was my attempt to cover the full table, including synchronous and asynchronous forms of each possibility of `intermediateResult`.